cpp/src/arrow/csv/reader_test.cc

```diff
@@ -0,0 +1,169 @@
+
+  std::vector<Future<std::shared_ptr<Table>>> task_futures(NTASKS);
+  for (int i = 0; i < NTASKS; i++) {
+    auto input = std::make_shared<io::BufferReader>(table_buffer);
+    ASSERT_OK_AND_ASSIGN(auto reader, reader_factory(input));
+    task_futures[i] = reader->ReadAsync();
+  }
+  auto combined_future = All(task_futures);
+  combined_future.Wait();
+
+  ASSERT_OK_AND_ASSIGN(std::vector<Result<std::shared_ptr<Table>>> results,
+                       combined_future.result());
+  for (auto&& result : results) {
+    ASSERT_OK_AND_ASSIGN(auto table, result);
+    ASSERT_EQ(NROWS, table->num_rows());
+  }
+}
+
+void TestNestedParallelism(
+    std::shared_ptr<internal::ThreadPool> thread_pool,
+    std::function<Result<std::shared_ptr<TableReader>>(std::shared_ptr<io::InputStream>)>
+        reader_factory) {
+  const int NROWS = 1000;
+  ASSERT_OK_AND_ASSIGN(auto table_buffer, MakeSampleCsvBuffer(NROWS));
+  auto input = std::make_shared<io::BufferReader>(table_buffer);
+  ASSERT_OK_AND_ASSIGN(auto reader, reader_factory(input));
+
+  Future<std::shared_ptr<Table>> table_future;
+
+  auto read_task = [&reader, &table_future]() mutable {
+    table_future = reader->ReadAsync();
+    return Status::OK();
+  };
+  ASSERT_OK_AND_ASSIGN(auto future, thread_pool->Submit(read_task));
+  ASSERT_TRUE(future.Wait(1));
+
+  if (future.is_finished()) {
+    ASSERT_TRUE(table_future.Wait(1));
+    if (table_future.is_finished()) {
+      ASSERT_OK_AND_ASSIGN(auto table, table_future.result());
+      ASSERT_EQ(table->num_rows(), NROWS);
+    }
+  }
+}  // namespace csv
+
+TEST(SerialReaderTests, Stress) {
+  auto task_factory = [](std::shared_ptr<io::InputStream> input_stream) {
+    return TableReader::Make(default_memory_pool(), io::AsyncContext(), input_stream,
+                             ReadOptions::Defaults(), ParseOptions::Defaults(),
+                             ConvertOptions::Defaults());
+  };
+  StressTableReader(task_factory);
+}
+
+TEST(SerialReaderTests, NestedParallelism) {
+  ASSERT_OK_AND_ASSIGN(auto thread_pool, internal::ThreadPool::Make(1));
+  auto task_factory = [](std::shared_ptr<io::InputStream> input_stream) {
+    return TableReader::Make(default_memory_pool(), io::AsyncContext(), input_stream,
+                             ReadOptions::Defaults(), ParseOptions::Defaults(),
+                             ConvertOptions::Defaults());
+  };
+  TestNestedParallelism(thread_pool, task_factory);
+}
+
+TEST(ThreadedReaderTests, Stress) {
+  ASSERT_OK_AND_ASSIGN(auto thread_pool, internal::ThreadPool::Make(1));
+  auto task_factory = [&thread_pool](std::shared_ptr<io::InputStream> input_stream)
+      -> Result<std::shared_ptr<TableReader>> {
+    ReadOptions read_options = ReadOptions::Defaults();
+    read_options.use_threads = true;
+    read_options.legacy_blocking_reads = true;
+    auto table_reader = TableReader::Make(
+        default_memory_pool(), io::AsyncContext(thread_pool.get()), input_stream,
```

Comment:
Hmm...if that is the purpose of `io::AsyncContext` then I have misused it.  I don't think that *should* be the purpose though.  It's possible I am wrong.  Although looking at the existing implementation I think I am indeed in misalignment...
```
// Default ReadAsync() implementation: simply issue the read on the context's executor
Future<std::shared_ptr<Buffer>> RandomAccessFile::ReadAsync(const AsyncContext& ctx, ...) {
  return DeferNotOk(ctx.executor->Submit(...);
}
```
...this logic doesn't work I don't think.  Well, it puts the burden on the user to transfer the continuation to the CPU context.  As it stands, if the user added a continuation, that continuation would run on the IO context.

When I see something like...

```
Future SomeIOOperation(AsyncContext ctx)
```

In my mind `ctx` is "This is the context on which work should resume when the I/O operation completes".  Indeed, it is the CPU context which has to be async most of the time.  There is typically not much harm in having a synchronous I/O context.

However, it sounds like the existing definition is "This is the context on which the I/O operation should run".  However, I think that should be owned by the FS.  Maybe AWS has a thread pool of its own that manages its operations.  Maybe a memory wrapping filesystem uses a single synchronous thread.  Maybe the FS is using libuv which has its own background thread pool.

Reply:
> However, it sounds like the existing definition is "This is the context on which the I/O operation should run". However, I think that should be owned by the FS.

Hmm, you may be right that the FS and/or stream implementation should own it. Can you open a JIRA for that?

> In my mind ctx is "This is the context on which work should resume when the I/O operation completes". 

That would be a weird API IMHO. You generally pass function parameters for what gets done in the function, not what happens when the function finishes executing.

Reply:
I can open a JIRA once we settle on what the desired behavior is but your second comment has left me unsure.  Are we saying that the method shouldn't accept an AsyncContext at all?  Otherwise, if it does, how should that be interpreted?

Do filesystems get to own their executors?  Or do caller's need to specify that in addition to the CPU executor?  Should these be specified every call?  Or should executors be somewhat "global" and specified through configuration (similar to how the default memory pool is set).